OrgFactory v2

cloudhead opened the pull request:
A new org factory that enables setup of orgs with ENS records in only two transactions, instead of five.

The steps that are combined into a single transaction that were previously multiple transactions are:
1. Create org
2. Register name
3. Set forward ENS record
4. Set reverse ENS record

The step that is still separate (and has to be) is the `commit` that happens before name registration.

cloudhead committed:
Add required methods to interfaces

cloudhead committed:
Add `OrgV2Factory`

Implements new methods for org creation in less steps.

cloudhead committed:
Add some 'require' calls

cloudhead committed:
Remove unused event

Ryanmtate commented on src/OrgV2Factory.sol:
```diff
@@ -0,0 +1,194 @@
+
+interface Resolver {
+    function multicall(bytes[] calldata data) external returns(bytes[] memory results);
+    function setAddr(bytes32, address) external;
```
Minor semantic change; do we want to add the names to the interface methods?

```solidity
function setAddr(bytes32 node, address addr) external;
```

https://docs.ens.domains/contract-api-reference/publicresolver#set-ethereum-address


I don't think it affects the method signature, since only parameter types are used in the signature.

cloudhead replied:
Yeah I was getting annoying warnings for some reason when I had the parameter names, due to a method with the same name existing... eg. `addr()` and `address addr`. That's why I removed them in some places, but not sure why the warning exists.

Ryanmtate replied:
Confirmed, annoying warnings! That's too bad.. not sure why that is as well. Possible workaround is prefixing all interface parameter names with an underscore. I've tried this just now and it seems to do the trick. Not sure if that is a solution for the underlying problem, though.

Something like this is workable:

```solidity
interface SafeFactory {
    function createProxy(address _masterCopy, bytes memory _data) external returns (Safe);
}

interface Resolver {
    function multicall(bytes[] calldata _data) external returns(bytes[] memory results);
    function setAddr(bytes32 _node, address _addr) external;
    function addr(bytes32 _node) external returns (address);
    function name(bytes32 _node) external returns (string memory);
}

interface Registrar {
    function commit(bytes32 _commitment) external;
    function register(string calldata _name, address _owner, uint256 _salt) external;
    function ens() external view returns (address);
    function radNode() external view returns (bytes32);
    function registrationFeeRad() external view returns (uint256);
    function minCommitmentAge() external view returns (uint256);
}

interface Safe {
    function setup(
        address[] calldata _owners,
        uint256 _threshold,
        address _to,
        bytes calldata _data,
        address _fallbackHandler,
        address _paymentToken,
        uint256 _payment,
        address payable _paymentReceiver
    ) external;

    function getThreshold() external returns (uint256);
    function isOwner(address _owner) external returns (bool);
}
```

cloudhead replied:
Yeah indeed, but it doesn't really matter, the interface is for internal use only and the parameter name only serves as documentation.

Ryanmtate commented on src/OrgV1.sol:
```diff
@@ -6,12 +6,15 @@ interface IERC20 {
 interface ENS {
-    function owner(bytes32 node) external view returns (address);
-    function resolver(bytes32 node) external view returns (address);
+    function owner(bytes32) external view returns (address);
```
https://docs.ens.domains/contract-api-reference/ens

Add method parameter names?

```solidity
function owner(bytes32 node) external view returns (address);
function resolver(bytes32 node) external view returns (address);
function setOwner(bytes32 node, address owner) external;
function setSubnodeOwner(bytes32 node, bytes32 label, address owner) external;
```


~~do we need to additional ENS method? e.g. add `ttl`?~~

Ryanmtate commented on src/OrgV2Factory.t.sol:
```diff
@@ -0,0 +1,125 @@
+
+interface Hevm {
+    function roll(uint256) external;
+    function store(address, bytes32, bytes32) external;
```
add parameter names?

```solidity
function roll(uint x) public external;
function store(address c, bytes32 loc, bytes32 val) public external;
```

Ryanmtate commented:
Nice work! Aside from some minor semantic parameter name additions, it looks good. Test cases are passing on my end.

Ryanmtate reviewed (dismissed):
Don't think the parameter name is required, otherwise the contracts look great!

cloudhead commented:
I just realized there is a potential front-running issue which the vanilla Registrar prevents, due to allowing `owner` to be specified after-the-fact in this contract. I'm going to see if there's a way to prevent that.

Ryanmtate commented:
> I just realized there is a potential front-running issue which the vanilla Registrar prevents, due to allowing `owner` to be specified after-the-fact in this contract. I'm going to see if there's a way to prevent that.

Are you referring to [#L89](https://github.com/radicle-dev/radicle-orgs/blob/cloudhead/org-v2-factory/src/OrgV2Factory.sol#L89) & [#L124](https://github.com/radicle-dev/radicle-orgs/blob/cloudhead/org-v2-factory/src/OrgV2Factory.sol#L124)?

It's not obvious to me what the implication of the front-running is here. Is there a reasonable way to test such a scenario?

sebastinez commented:
Looks great to me! Tests are also passing on my end.
Liking the new flow of allowing the setting of the forward and reverse record in the same tx! Kudos :clap: 
Regarding the OrgV1 contract, I think the `executeTransaction` method would be great for now, as written on Discord.
Eventually I'm thinking we could give the contract more functionality, and expand them from only doing anchoring.
For example we could expand the contribution rewards functionality to each org, and use them also for on-chain proof verification..
But this will need some more thinking.. :thinking:

sebastinez reviewed (dismissed):
:100:

rudolfs reviewed (commented):
Looks good from what I can tell. Looking forward to the more integrated org creation flow!

CodeSandwich reviewed (dismissed):
LGTM

cloudhead committed:
Prevent front-running of registrations

Without this change it's possible to front-run the registration
step by setting any 'owner'.

cloudhead committed:
Factor out commitment code

cloudhead committed:
Add ability to register and reclaim name

cloudhead commented:
> > I just realized there is a potential front-running issue which the vanilla Registrar prevents, due to allowing `owner` to be specified after-the-fact in this contract. I'm going to see if there's a way to prevent that.
> 
> Are you referring to [#L89](https://github.com/radicle-dev/radicle-orgs/blob/cloudhead/org-v2-factory/src/OrgV2Factory.sol#L89) & [#L124](https://github.com/radicle-dev/radicle-orgs/blob/cloudhead/org-v2-factory/src/OrgV2Factory.sol#L124)?
> 
> It's not obvious to me what the implication of the front-running is here. Is there a reasonable way to test such a scenario?

Yeah, once a name is committed to, it's possible to register with any owner if you know the name and salt. So once the registrant submits the registration tx, someone could front-run it and put their own address.

I pushed a fix which creates a commitment to the final address, so this should no longer be an issue.

cloudhead committed:
Make user pay for commitment
